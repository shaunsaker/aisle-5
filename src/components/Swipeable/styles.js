import { StyleSheet } from 'react-native';

import styleConstants from '../../styleConstants';

const styles = StyleSheet.create({
  container: {},
});

export default styles;
