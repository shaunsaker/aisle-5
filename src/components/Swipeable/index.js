import React from 'react';
import PropTypes from 'prop-types';
import { View, PanResponder, Animated } from 'react-native';

import styleConstants from '../../styleConstants';

import styles from './styles';

export default class Swipeable extends React.Component {
  constructor(props) {
    super(props);

    this.gestureThreshold = 35;
    this.gestureDelay = this.gestureThreshold * -1;
    this.snapStartDuration = 150;
    this.snapEndDuration = 300; // when item is removed

    this.state = {
      pan: new Animated.ValueXY(),
    };
  }

  static get propTypes() {
    return {
      width: PropTypes.number,
      onSwipeStart: PropTypes.func,
      onSwipeEnd: PropTypes.func,
      onSwiped: PropTypes.func,
      children: PropTypes.node,
    };
  }

  componentWillMount() {
    const { pan } = this.state;
    const { width, onSwipeStart, onSwipeEnd, onSwiped } = this.props;

    this.panResponder = PanResponder.create({
      onStartShouldSetPanResponder: () => false,

      onMoveShouldSetPanResponder: () => true,

      onPanResponderTerminationRequest: () => false,

      onPanResponderMove: (event, gestureState) => {
        if (gestureState.dx > this.gestureThreshold) {
          const newX = gestureState.dx + this.gestureDelay;

          pan.setValue({ x: newX, y: 0 });

          if (onSwipeStart) {
            onSwipeStart();
          }
        }
      },

      onPanResponderRelease: (event, gestureState) => {
        if (gestureState.dx < width / 2) {
          Animated.timing(pan, {
            toValue: { x: 0, y: 0 },
            duration: this.snapStartDuration,
            easing: styleConstants.easing,
          }).start(() => {
            if (onSwipeEnd) {
              onSwipeEnd();
            }
          });
        } else {
          Animated.timing(pan, {
            toValue: { x: width, y: 0 },
            duration: this.snapEndDuration,
            easing: styleConstants.easing,
          }).start(() => {
            if (onSwipeEnd) {
              onSwipeEnd();
            }

            if (onSwiped) {
              onSwiped();
            }
          });
        }
      },
    });
  }

  render() {
    const { pan } = this.state;
    const { width, children } = this.props;
    const color = pan.x.interpolate({
      inputRange: [0, width],
      outputRange: ['rgba(255, 0, 0, 0)', 'rgba(255, 0, 0, 1)'],
    });

    return (
      <View style={styles.container}>
        <Animated.View
          {...this.panResponder.panHandlers}
          style={[pan.getLayout(), { backgroundColor: color }]}
        >
          {children}
        </Animated.View>
      </View>
    );
  }
}
