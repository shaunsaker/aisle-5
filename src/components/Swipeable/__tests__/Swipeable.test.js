import React from 'react';
import renderer from 'react-test-renderer';

import Swipeable from '..';

describe('Swipeable', () => {
  const width = 100;

  describe('renders', () => {
    it('renders with minimum required props', () => {
      const component = renderer.create(<Swipeable width={width} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe('lifecycle methods', () => {
    // FIXME: How to test these event listeners
  });
});
