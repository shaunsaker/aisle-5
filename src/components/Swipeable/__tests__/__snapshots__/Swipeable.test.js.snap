// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swipeable renders renders with minimum required props 1`] = `
<View
  style={Object {}}
>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(255, 0, 0, 0)",
        "left": 0,
        "top": 0,
      }
    }
  />
</View>
`;
